### PR TITLE
[Core] Optimize kv-cache transfer with layer-wise pipeline

### PR DESCRIPTION
**Pull request by ZeldaHuang**
1. Add args `migration_num_layers`, transfer kv-cache with every `migration_num_layers`, ~align the migration cache tensor shape with the VLLM cache shape~.
2. Add gloo bf16 support.

### CODE REVIEW

**Comment by ZeldaHuang** [comment]
`model`=llama7b
`migration_num_layers`=1
`block_size`=16


backend gloo
| transfer blocks | before | after |
|-------|-------|-------|
| 8  |31.9ms|16.5ms  |
| 16  |48.5ms| 26.3ms  |
| 32  |110.1ms| 58ms  |

backend nccl
| transfer blocks | before | after |
|-------|-------|-------|
| 8  |10.4ms| 6.5ms|
| 16  |17.3ms| 8.3ms |
| 32  |30.6ms|  18.5ms|

File: llumnix/backends/vllm/migrate_backend.py [hunk 1]
```diff
@@ -85,48 +86,41 @@ def destory_col(self) -> None:
...
-            ray_obj = self.actor.exec_method.remote(self.is_driver_worker, src_handle, "do_send", None, send_blocks)
+            ray_obj = self.actor.exec_method.remote(self.is_driver_worker, src_handle, "do_send", send_blocks)
             if rpc_numpy_cache is not None:
                 self.do_recv(rpc_numpy_cache, recv_blocks)
             rpc_numpy_cache = ray.get(ray_obj)
             recv_blocks = dst_blocks[start_idx:start_idx+offset]
         self.do_recv(rpc_numpy_cache, recv_blocks)
 
-    def do_send(self, dst_handle, blocks: List[int]):
+    def do_send(self, blocks: List[int]):
```
Comment (KuilongCui) [inline llumnix/backends/vllm/migrate_backend.py:104]: Although dst_handle is not used, retain it to maintain compatibility with the MigrateBackendBase interface.

Comment (ZeldaHuang) [reply llumnix/backends/vllm/migrate_backend.py:104]: In `RPCMigrateBackend`, `def do_recv(self, rpc_numpy_cache, blocks: List[int])` doesn't follow definition in `MigrateBackendBase`,  It's not a clear api, we only use `src_handle` and `dst_handle` in `RayMigrateBackend`.

Comment (KuilongCui) [reply llumnix/backends/vllm/migrate_backend.py:104]: how about  revising the interface in MigrateBackendBase ?

**Pull request merged**